submissions/antonkottans/tjsw-post-oop/index.js

```diff
@@ -0,0 +1,165 @@
+const KEYS_VALUE_SEPARATOR = ": ";
+const PROPS_SEPARATOR = "; ";
+
+class Creature {
+  constructor({ name, gender, friends = [], species, saying }) {
+    this.name = name;
+    this.species = species;
+    this.gender = gender;
+    this.friends = friends;
+    this.saying = saying;
+  }
+
+  createArrayOfProps() {
+    return [
+      this.name,
+      this.species,
+      this.gender,
+      this.saying,
+      this.friendsToString(),
+    ];
+  }
+
+  friendsToString() {
+    return this.friends.length
+      ? this.friends.map((friend) => friend.name).join(", ")
+      : "no friends";
+  }
+
+  addFriends(...newFriends) {
+    return (this.friends = [...this.friends, ...newFriends]);
+  }
+}
+
+class Animal extends Creature {
+  constructor({ name, gender, friends, paws, species, saying }) {
+    super({ name, gender, friends, species, saying });
+    this.paws = paws;
+  }
+
+  createArrayOfProps() {
+    return [...super.createArrayOfProps(), this.paws];
+  }
+}
+
+class AnimalWithTail extends Animal {
+  constructor({ name, gender, friends, paws, species, saying, tail }) {
+    super({ name, gender, friends, paws, species, saying });
+    this.tail = tail;
+  }
+
+  createArrayOfProps() {
+    return [...super.createArrayOfProps(), this.tail];
+  }
+}
+
+class Human extends Creature {
+  constructor({ name, gender, friends, legs, hands, saying }) {
+    super({ name, gender, friends, species: "human", saying });
+    this.legs = legs;
+    this.hands = hands;
+  }
+
+  createArrayOfProps() {
+    return [...super.createArrayOfProps(), this.legs, this.hands];
+  }
+}
+
+class Dog extends AnimalWithTail {
+  constructor({ name, gender, friends, paws, tail = "ordinary tail", saying }) {
+    super({ name, gender, friends, paws, tail, species: "dog", saying });
+  }
+}
+
+class Cat extends AnimalWithTail {
+  constructor({
+    name,
+    gender,
+    friends,
+    paws,
+    tail = "ordinary tail",
+    saying = "meow!",
+  }) {
+    super({ name, gender, friends, paws, tail, species: "cat", saying });
+    this.saying = saying;
```

Comment:
Any good reasons to initialize saying two ways?

Reply:
i added this.saying = saying; for catwoman. there is no need for that

Reply:
@AntonKottans You are doing good and will be fine. There is no learning without an effort.